src/locale/keymap-util.c

```diff
@@ -780,3 +782,228 @@ int x11_convert_to_vconsole(Context *c) {
+                return false;
+        }
+        return true;
+#else
+        return false;
+#endif
+}
+
+#if HAVE_LOCALEGEN
+static bool locale_encoding_is_utf8_or_unspecified(const char *locale) {
+        const char *c = strchr(locale, '.');
+        return !c || strcaseeq(c,  ".UTF-8") || strcasestr(locale, ".UTF-8@");
+}
+
+static int locale_gen_entry_check_supported(const char *locale_entry) {
+        /* Returns an error valus <= 0 if the locale-gen entry is invalid or unsupported,
+         * 1 in case the locale entry is valid, and -EOPNOTSUPP specifically in case
+         * the distributor has not provided us with a SUPPORTED file to check
+         * locale for validity. */
+
+        _cleanup_fclose_ FILE *f = NULL;
+        int r;
+
+        assert(locale_entry);
+
+        /* Locale templates without country code are never supported */
+        if (!strstr(locale_entry, "_"))
+                return -EINVAL;
+
+        if (access("/usr/share/i18n/SUPPORTED", F_OK) < 0) {
+                if (errno != ENOENT)
+                        log_warning_errno(errno, "Unable to determine whether /usr/share/i18n/SUPPORTED exists: %m");
+                else
+                        log_warning("Can not check validity of locale entry %s: /usr/share/i18n/SUPPORTED does not exist", locale_entry);
+                return -EOPNOTSUPP;
+        }
+
+        f = fopen("/usr/share/i18n/SUPPORTED", "re");
+        if (!f)
+                return -errno;
+
+        for (;;) {
+                _cleanup_free_ char *line = NULL;
+
+                r = read_line(f, LONG_LINE_MAX, &line);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                line = strstrip(line);
+                if (strcasecmp_ptr(line, locale_entry) == 0)
+                        return 1;
+        }
+
+        /* if we are here, the locale is unsupported or invalid */
+        return -EINVAL;
+}
+#endif
+
+int locale_gen_enable_locale(const char *locale) {
+#if HAVE_LOCALEGEN
+        _cleanup_fclose_ FILE *fr = NULL, *fw = NULL;
+        _cleanup_(unlink_and_freep) char *temp_path = NULL;
+        _cleanup_free_ char *locale_entry = NULL;
+        bool locale_enabled = false, first_line = false;
+        struct stat st;
+        int r;
+
+        if (isempty(locale))
+                return 0;
+
+        if (locale_encoding_is_utf8_or_unspecified(locale)) {
+                locale_entry = strjoin(locale, " UTF-8");
+                if (!locale_entry)
+                        return -ENOMEM;
+        } else
+                return -ENOEXEC; /* We do not process non-UTF-8 locale */
+
+        r = locale_gen_entry_check_supported(locale_entry);
+        if ((r <= 0) && (r != -EOPNOTSUPP))
+                return r;
+
+        fr = fopen("/etc/locale.gen", "re");
+        if (!fr)
+                return -errno;
```

Comment:
I think that if the file doesn't exist, we should just write the new file. After all, you append new locales at the end if they are not found in the file, and a nonexistent file is almost the same as an empty file which is almost the same without some specific entry. We generally expect that `/etc` can be emptied and tools will DTRT.